mistletoe/markdown_renderer.py

```diff
@@ -0,0 +1,544 @@
+            yield Fragment(dest_part)
+            if token.title:
+                yield from (
+                    Fragment(" ", wordwrap=True),
+                    Fragment(token.title_delimiter),
+                    Fragment(token.title, wordwrap=True),
+                    Fragment(
+                        ")" if token.title_delimiter == "(" else token.title_delimiter,
+                    ),
+                )
+            yield Fragment(")")
+        elif token.dest_type == "full":
+            # "[" description "][" label "]"
+            yield from (
+                Fragment("["),
+                Fragment(token.label, wordwrap=True),
+                Fragment("]"),
+            )
+        elif token.dest_type == "collapsed":
+            # "[" description "][]"
+            yield Fragment("[]")
+        else:
+            # "[" description "]"
+            pass
+
+    def render_auto_link(self, token: span_token.AutoLink) -> Iterable[Fragment]:
+        yield Fragment("<" + token.children[0].content + ">")
+
+    def render_escape_sequence(
+        self, token: span_token.EscapeSequence
+    ) -> Iterable[Fragment]:
+        yield Fragment("\\" + token.children[0].content)
+
+    def render_line_break(self, token: span_token.LineBreak) -> Iterable[Fragment]:
+        yield Fragment(
+            token.content + "\n", wordwrap=token.soft, hard_line_break=not token.soft
+        )
+
+    def render_html_span(self, token: span_token.HTMLSpan) -> Iterable[Fragment]:
+        yield Fragment(token.content)
+
+    def render_link_reference_definition(
+        self, token: LinkReferenceDefinition
+    ) -> Iterable[Fragment]:
+        yield from (
+            Fragment("["),
+            Fragment(token.label, wordwrap=True),
+            Fragment("]: ", wordwrap=True),
+            Fragment(
+                "<" + token.dest + ">"
+                if token.dest_type == "angle_uri"
+                else token.dest,
+            ),
+        )
+        if token.title:
+            yield from (
+                Fragment(" ", wordwrap=True),
+                Fragment(token.title_delimiter),
+                Fragment(token.title, wordwrap=True),
+                Fragment(
+                    ")" if token.title_delimiter == "(" else token.title_delimiter,
+                ),
+            )
+
+    # rendering of block tokens.
+    # rendered into sequences of lines (strings), to be joined by newlines.
+
+    def render_document(
+        self, token: block_token.Document, max_line_length: int
+    ) -> Iterable[str]:
+        return self.blocks_to_lines(token.children, max_line_length=max_line_length)
+
+    def render_heading(
+        self, token: block_token.Heading, max_line_length: int
+    ) -> Iterable[str]:
+        # note: no word wrapping, because atx headings always fit on a single line.
+        line = "#" * token.level
+        text = next(self.span_to_lines(token.children, max_line_length=None), "")
```

Comment:
OK, even if I still think we could omit the passing of `max_line_length` all the way down when we've got it stored in `self` now (and specifically handle just the extra cases where wrapping is not allowed - like in here). But OK, good enough for now, we can refactor later on if there will be time. ;)

Reply:
The reason for the parameter to still be there in the method calls is that the effective line length changes during rendering. For example, if you're rendering a block quote, then each line is prefixed with a `> `, and the effective max line length decreases by two. So while the naming can be discussed, it's not really possible to use the value stored with `self`.

Reply:
Oh, I see, makes sense. This reminds me of I haven't done a full review of all the code yet. :') I'm afraid I don't have enough capacity for that many lines now, so I guess I will just risk it and merge it without a detailed review.